### PR TITLE
CAL-98 Added UDP Stream generator to Alliance SDK and created a mvn profile

### PR DESCRIPTION
**Pull request by bdeining**
#### What does this PR do?
This PR adds a UDP Stream generator to Alliance SDK for testing purposes.
#### Who is reviewing it?
@glenhein @dcruver @jlcsmith @kcwire 
#### How should this be tested?
Build and Install Alliance, configure a UDP Stream, and follow the README in `sample-mpegts-stream-generator` to ingest a UDP stream
#### What are the relevant tickets?
https://codice.atlassian.net/browse/CAL-98
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                        }
+                    });
+
+            Channel ch = bootstrap.bind(0)
+                    .sync()
+                    .channel();
+
+            long bytesSent = 0;
+
+            long tsPacketCount = 618570;
```
Comment (glenhein) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: the packet count can be calculated by dividing the file size by 188.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            long tsDurationMillis = (4 * 60 + 39) * 1000;
```
Comment (glenhein) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Can we depend on ffmpeg? You could use ffmpeg to get the duration, that way users don't have to tweak the source code for every file they want to transmit.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                    }
+                }
+            }
+
+            long endTime = System.currentTimeMillis();
+
+            LOGGER.info("elapsed: {}", endTime - startTime);
+
+            if (!ch.closeFuture()
+                    .await(5000)) {
```
Comment (glenhein) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: This probably doesn't need to be 5 seconds, it's just waiting for the packets to be sent, 100ms is probably plenty of time.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+            try (final FileInputStream fis = new FileInputStream(videoFilePath)) {
+
+                byte[] buffer = new byte[188];
+                int c;
+                while ((c = fis.read(buffer)) != -1) {
+                    bytesSent += c;
+
+                    ChannelFuture cf = ch.writeAndFlush(new DatagramPacket(Unpooled.copiedBuffer(
+                            buffer), new InetSocketAddress("127.0.0.1", 50000)));
```
Comment (glenhein) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Should the IP address and port number be made into command line args with sensible defaults?

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 3 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Changed

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 4 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Made IP and Port configurable

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 2 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Removed hard coded duration and added a package to get the video duration

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 4 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Also added defaults

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 2 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Used ffmpeg to get the video duration

**Comment by glenhein** [comment]
ben has one small default value to change, but otherwise, it builds and runs as expected, looks great!

**Comment by bdeining** [comment]
@glenhein changed default value to 50000

File: distribution/sdk/pom.xml [hunk 5]
```diff
@@ -31,6 +31,7 @@
     </properties>
 
     <modules>
+        <module>sample-mpegts-stream-generator</module>
```
Comment (jlcsmith) [inline distribution/sdk/pom.xml]: nit - should be sample-mpegts-streamgenerator

**Comment by jlcsmith** [comment]
LGTM 👍

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+    private static final String DEFAULT_IP = "127.0.0.1";
+
+    private static final int DEFAULT_PORT = 50000;
+
+    private static final String SUPPRESS_PRINTING_BANNER_FLAG = "-hide_banner";
+
+    private static final String INPUT_FILE_FLAG = "-i";
+
+    private static final boolean DONT_HANDLE_QUOTING = false;
```
Comment (lcrosenbu) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Recommend changing this to `HANDLE_QUOTING`. 
Don't not avoid double negative logic. I get too confused. :)

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        String[] arguments = args[0].split(",");
+
+        if (arguments.length < 1) {
+            LOGGER.error("Unable to start stream : no arguments specified.");
+            return;
+        }
+
+        String videoFilePath = arguments[0];
+        if (StringUtils.isBlank(videoFilePath)) {
+            LOGGER.error("Unable to start stream : no video file path specified.");
```
Comment (lcrosenbu) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Recommend pulling input error logging out to a helper method, with more robust usage description.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+            } catch (NumberFormatException e) {
+                LOGGER.error("Unable to parse specified port : {}.  Using Default : {}",
+                        arguments[2],
+                        DEFAULT_PORT);
+                port = DEFAULT_PORT;
+            }
+        }
+
+        if (StringUtils.isBlank(ip)) {
+            LOGGER.error("Unable to start stream : no ip address sspecified.");
```
Comment (lcrosenbu) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Recommend pulling input error logging out to a helper method, with more robust usage description.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        if (SystemUtils.IS_OS_LINUX) {
+            return "linux/ffmpeg";
+        } else if (SystemUtils.IS_OS_MAC) {
+            return "osx/ffmpeg";
+        } else if (SystemUtils.IS_OS_SOLARIS) {
+            return "solaris/ffmpeg";
+        } else if (SystemUtils.IS_OS_WINDOWS) {
+            return "windows/ffmpeg.exe";
+        } else {
+            throw new RuntimeException("OS is not Linux, Mac, Solaris, or Windows."
```
Comment (lcrosenbu) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Recommend moving all input validation, to include OS compatibility, to the main() area, and include requirements with usage statement.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+                    + " No FFmpeg binary is available for this OS, so the plugin will not work.");
+        }
+    }
+
+    private static Duration getVideoDuration(final String videoFilePath)
+            throws IOException, InterruptedException {
+        final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
Comment (lcrosenbu) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Recommend try with resources for I/O

File: distribution/sdk/pom.xml (hunk 5 above)
Comment (bdeining) [inline distribution/sdk/pom.xml]: Renamed module

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 6 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Renamed

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 7 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Done

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 8 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Done

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 9 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: This isn't really input validation;  it's used for the ffmpeg command called from main.  I've cleaned up the error messages.

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 10 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: This I/O stuff doesn't implement AutoCloseable, so we can't use try-with-resources, but i've refactored the try catches to a centralized location and improved the error messages.

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        } else if (SystemUtils.IS_OS_WINDOWS) {
+            return "windows/ffmpeg.exe";
+        } else {
+            throw new IllegalStateException("OS is not Linux, Mac, Solaris, or Windows."
+                    + " No FFmpeg binary is available for this OS, so the plugin will not work.");
+        }
+    }
+
+    private static Duration getVideoDuration(final String videoFilePath) {
+        final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
Comment (coyotesqrl) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: While `close()` on BAOS is a noop, FindBugs (and I believe most other static analysis tools) aren't quite smart enough to figure that out. Using try-with-resources here is more about quieting those reports. Also, it's just generally good practice to treat every `Closeable` the same way so it becomes habitual for everyone touching the codebase.

Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Done

File: distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 10 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-stream-generator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Disregard, didn't realize you could do this

File: distribution/sdk/sample-mpegts-streamgenerator/pom.xml [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        <artifactId>sdk</artifactId>
+        <version>0.2-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <name>Alliance :: Distribution :: Video :: Mpegts Stream Generator</name>
+    <artifactId>sample-mpegts-streamgenerator</artifactId>
+    <packaging>bundle</packaging>
+    <properties>
+        <ffmpeg.version>2.8.4_2</ffmpeg.version>
```
Comment (coyotesqrl) [inline distribution/sdk/sample-mpegts-streamgenerator/pom.xml]: Why this old version of ffmpeg?

The OWASP-passing current version is 3.1.1_1. (https://github.com/codice/thirdparty/blob/master/ffmpeg/pom.xml#L14)

Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/pom.xml]: Fixed

File: distribution/sdk/sample-mpegts-streamgenerator/pom.xml [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    <parent>
+        <groupId>org.codice.alliance.distribution</groupId>
+        <artifactId>sdk</artifactId>
+        <version>0.2-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <name>Alliance :: Distribution :: Video :: Mpegts Stream Generator</name>
+    <artifactId>sample-mpegts-streamgenerator</artifactId>
+    <packaging>bundle</packaging>
```
Comment (pklinef) [inline distribution/sdk/sample-mpegts-streamgenerator/pom.xml]: Is this really a bundle that will run in the container?  If not I would suggest changing this to jar and using shade to embed dependencies.

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+
+            LOGGER.info("Video Duration : {}", tsDurationMillis);
+
+            double delayPerPacket = (double) tsDurationMillis / (double) tsPacketCount;
+
+            long startTime = System.currentTimeMillis();
+
+            int packetsSent = 0;
+
+            try (final FileInputStream fis = new FileInputStream(videoFile)) {
```
Comment (coyotesqrl) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Wrap this in a `BufferedInputStream` so your calls to `read` aren't literally pushing the pointer one byte at a time through the file. Much faster/more efficient.

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+                    bytesSent += c;
+
+                    ChannelFuture cf = ch.writeAndFlush(new DatagramPacket(Unpooled.copiedBuffer(
+                            buffer), new InetSocketAddress(ip, port)));
+
+                    cf.await();
+
+                    packetsSent++;
+
+                    if (packetsSent % 100 == 0) {
```
Comment (coyotesqrl) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Any particular reason you're modding on 100 here, or just an arbitrary number of packets between delays?

File: distribution/sdk/sample-mpegts-streamgenerator/pom.xml [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+            <artifactId>commons-exec</artifactId>
+            <version>1.2</version>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
```
Comment (pklinef) [inline distribution/sdk/sample-mpegts-streamgenerator/pom.xml]: This project only runs from mvn and does not use OSGi, remove this plugin and change the packaging to jar.

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+            return null;
+        } catch (IOException | IllegalStateException e) {
+            logErrorMessage(String.format("Unable to execute ffmpeg command. %s", e), false);
+            return null;
+        }
+    }
+
+    private static Duration parseVideoDuration(final String ffmpegOutput)
+            throws IllegalArgumentException {
+        final Pattern pattern = Pattern.compile("Duration: \\d\\d:\\d\\d:\\d\\d\\.\\d+");
```
Comment (coyotesqrl) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: I know this is a ridiculous question, but do we not support videos greater than 100 hours in duration or is such a beast just not possible?

**Comment by coyotesqrl** [comment]
Phil's and my comments aside, this lgtm.

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 15 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: This is just an arbitrary number of packets

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 17 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: While 100 hours would be possible to stream, it's unlikely that an FMV clip will be this long

File: distribution/sdk/sample-mpegts-streamgenerator/pom.xml (hunk 13 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/pom.xml]: Changed to jar

File: distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java (hunk 14 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/src/main/java/org/codice/alliance/distribution/sdk/video/stream/mpegts/MpegTsUdpClient.java]: Wrapped in `BufferedInputStream`

File: distribution/sdk/sample-mpegts-streamgenerator/pom.xml (hunk 16 above)
Comment (bdeining) [inline distribution/sdk/sample-mpegts-streamgenerator/pom.xml]: Done

**Comment by pklinef** [comment]
👍

**Pull request merged**